File: android/src/main/kotlin/com/umda/barcode_scanner_pro/ScannerConfig.kt
package com.umda.barcode_scanner_pro

import android.util.Size
import androidx.camera.core.CameraSelector

/** Normalized scan-area rectangle in [0,1] preview space. */
internal data class ScanAreaConfig(
    val left: Double,
    val top: Double,
    val width: Double,
    val height: Double,
) {
    val isFull: Boolean get() = left == 0.0 && top == 0.0 && width == 1.0 && height == 1.0
}

/**
 * Parsed, type-safe representation of the configuration map sent from Dart
 * (`ScannerConfiguration.toMap`). Keeps channel-key strings confined to one place.
 */
internal data class ScannerConfig(
    val cameraFacing: Int,
    val resolution: Int,
    val formatMask: Int,
    val scanMode: Int,
    val scanArea: ScanAreaConfig,
    val continuousScanning: Boolean,
    val duplicateTimeoutMs: Long,
    val enableAutoFocus: Boolean,
    val enableAutoZoom: Boolean,
    val enableSound: Boolean,
    val enableVibration: Boolean,
    val enablePinchZoom: Boolean,
    val enableTapFocus: Boolean,
    val frameRateLimit: Int,
    val returnImage: Boolean,
    val detectInverted: Boolean,
) {
    val lensFacing: Int
        get() = if (cameraFacing == 1) CameraSelector.LENS_FACING_FRONT
        else CameraSelector.LENS_FACING_BACK

    /** Target analysis resolution; CameraX picks the nearest supported size. */
    val targetResolution: Size
        get() = when (resolution) {
            0 -> Size(480, 640)
            1 -> Size(720, 1280)
            2 -> Size(1080, 1920)
            else -> Size(1080, 1920)
        }

    /** Minimum interval between decoded frames, derived from the FPS cap. */
    val frameIntervalMs: Long get() = (1000L / frameRateLimit.coerceAtLeast(1))

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromMap(map: Map<String, Any?>): ScannerConfig {
            val area = map["scanArea"] as? Map<String, Any?> ?: emptyMap()
            fun d(v: Any?, def: Double) = (v as? Number)?.toDouble() ?: def
            return ScannerConfig(
                cameraFacing = (map["camera"] as? Number)?.toInt() ?: 0,
                resolution = (map["resolution"] as? Number)?.toInt() ?: 1,
                formatMask = (map["formats"] as? Number)?.toInt() ?: 0,
                scanMode = (map["scanMode"] as? Number)?.toInt() ?: 1,
                scanArea = ScanAreaConfig(
                    d(area["left"], 0.0), d(area["top"], 0.0),
                    d(area["width"], 1.0), d(area["height"], 1.0),
                ),
                continuousScanning = map["continuousScanning"] as? Boolean ?: true,
                duplicateTimeoutMs = (map["duplicateTimeoutMs"] as? Number)?.toLong() ?: 1000L,
                enableAutoFocus = map["enableAutoFocus"] as? Boolean ?: true,
                enableAutoZoom = map["enableAutoZoom"] as? Boolean ?: false,
                enableSound = map["enableSound"] as? Boolean ?: true,
                enableVibration = map["enableVibration"] as? Boolean ?: true,
                enablePinchZoom = map["enablePinchZoom"] as? Boolean ?: true,
                enableTapFocus = map["enableTapFocus"] as? Boolean ?: true,
                frameRateLimit = (map["frameRateLimit"] as? Number)?.toInt() ?: 15,
                returnImage = map["returnImage"] as? Boolean ?: false,
                detectInverted = map["detectInverted"] as? Boolean ?: false,
            )
        }
    }
}

File: ios/Classes/ScannerConfig.swift
import AVFoundation
import Foundation

/// Normalized scan-area rectangle in [0,1] preview space.
struct ScanAreaConfig {
  let left: Double
  let top: Double
  let width: Double
  let height: Double

  var isFull: Bool { left == 0 && top == 0 && width == 1 && height == 1 }
}

/// Type-safe representation of the configuration sent from Dart
/// (`ScannerConfiguration.toMap`).
struct ScannerConfig {
  let cameraFacing: Int
  let resolution: Int
  let formatMask: Int
  let scanMode: Int
  let scanArea: ScanAreaConfig
  let continuousScanning: Bool
  let duplicateTimeoutMs: Double
  let enableAutoFocus: Bool
  let enableSound: Bool
  let enableVibration: Bool
  let frameRateLimit: Int
  let returnImage: Bool
  let detectInverted: Bool

  var devicePosition: AVCaptureDevice.Position {
    cameraFacing == 1 ? .front : .back
  }

  /// Session preset mapped from the resolution enum.
  var sessionPreset: AVCaptureSession.Preset {
    switch resolution {
    case 0: return .vga640x480
    case 1: return .hd1280x720
    case 2: return .hd1920x1080
    default: return .hd1920x1080
    }
  }

  /// Minimum interval (seconds) between decoded frames from the FPS cap.
  var frameInterval: CFTimeInterval {
    1.0 / Double(max(frameRateLimit, 1))
  }

  static func from(_ map: [String: Any?]) -> ScannerConfig {
    let area = map["scanArea"] as? [String: Any?] ?? [:]
    func d(_ v: Any??, _ def: Double) -> Double { (v as? NSNumber)?.doubleValue ?? def }
    func i(_ v: Any??, _ def: Int) -> Int { (v as? NSNumber)?.intValue ?? def }
    func b(_ v: Any??, _ def: Bool) -> Bool { (v as? NSNumber)?.boolValue ?? def }
    return ScannerConfig(
      cameraFacing: i(map["camera"], 0),
      resolution: i(map["resolution"], 1),
      formatMask: i(map["formats"], 0),
      scanMode: i(map["scanMode"], 1),
      scanArea: ScanAreaConfig(
        left: d(area["left"], 0), top: d(area["top"], 0),
        width: d(area["width"], 1), height: d(area["height"], 1)
      ),
      continuousScanning: b(map["continuousScanning"], true),
      duplicateTimeoutMs: d(map["duplicateTimeoutMs"], 1000),
      enableAutoFocus: b(map["enableAutoFocus"], true),
      enableSound: b(map["enableSound"], true),
      enableVibration: b(map["enableVibration"], true),
      frameRateLimit: i(map["frameRateLimit"], 15),
      returnImage: b(map["returnImage"], false),
      detectInverted: b(map["detectInverted"], false)
    )
  }
}
